File: solutions/elixir/proverb/1/lib/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """
  @spec recite(strings :: [String.t()]) :: String.t()
  def recite([]), do: ""

  def recite([first | rest]) do
    recite_helper([first | rest]) <> "And all for the want of a #{first}.\n"
  end

  defp recite_helper(list) when length(list) <= 1, do: ""

  defp recite_helper([first | [second | rest]]) do
    "For want of a #{first} the #{second} was lost.\n" <> recite_helper([second | rest])
  end
end

File: solutions/ruby/proverb/1/proverb.rb
class Proverb
  attr_reader :words, :qualifier

  def initialize(*words, qualifier: nil)
    @words = words
    @qualifier = qualifier
  end

  def to_s
    return '' if @words.empty?

    lines = words.each_cons(2).map do |want, lost|
      "For want of a #{want} the #{lost} was lost."
    end

    qualifier_text = qualifier ? "#{qualifier} " : ''
    lines << "And all for the want of a #{qualifier_text}#{words[0]}."

    lines.join("\n")
  end
end
